File: lib/js/bs_SetM.js
'use strict';

var Bs_Sort = require("./bs_Sort.js");
var Bs_internalAVLset = require("./bs_internalAVLset.js");

function removeMutateAux(nt, x, cmp) {
  var k = nt.key;
  var c = cmp(x, k);
  if (c) {
    if (c < 0) {
      var match = nt.left;
      if (match !== null) {
        nt.left = removeMutateAux(match, x, cmp);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return nt;
      }
    } else {
      var match$1 = nt.right;
      if (match$1 !== null) {
        nt.right = removeMutateAux(match$1, x, cmp);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return nt;
      }
    }
  } else {
    var l = nt.left;
    var r = nt.right;
    if (l !== null) {
      if (r !== null) {
        nt.right = Bs_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else if (r !== null) {
      return r;
    } else {
      return l;
    }
  }
}

function remove(d, v) {
  var oldRoot = d.data;
  var M = d.dict;
  if (oldRoot !== null) {
    var newRoot = removeMutateAux(oldRoot, v, M[/* cmp */0]);
    if (newRoot !== oldRoot) {
      d.data = newRoot;
      return /* () */0;
    } else {
      return 0;
    }
  } else {
    return /* () */0;
  }
}

function removeArrayMutateAux(_t, xs, _i, len, cmp) {
  while(true) {
    var i = _i;
    var t = _t;
    if (i < len) {
      var ele = xs[i];
      var u = removeMutateAux(t, ele, cmp);
      if (u !== null) {
        _i = i + 1 | 0;
        _t = u;
        continue ;
        
      } else {
        return Bs_internalAVLset.empty0;
      }
    } else {
      return t;
    }
  };
}

function removeMany(d, xs) {
  var oldRoot = d.data;
  if (oldRoot !== null) {
    var len = xs.length;
    var M = d.dict;
    var newRoot = removeArrayMutateAux(oldRoot, xs, 0, len, M[/* cmp */0]);
    if (newRoot !== oldRoot) {
      d.data = newRoot;
      return /* () */0;
    } else {
      return 0;
    }
  } else {
    return /* () */0;
  }
}

function removeMutateCheckAux(nt, x, removed, cmp) {
  var k = nt.key;
  var c = cmp(x, k);
  if (c) {
    if (c < 0) {
      var match = nt.left;
      if (match !== null) {
        nt.left = removeMutateCheckAux(match, x, removed, cmp);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return nt;
      }
    } else {
      var match$1 = nt.right;
      if (match$1 !== null) {
        nt.right = removeMutateCheckAux(match$1, x, removed, cmp);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return nt;
      }
    }
  } else {
    removed[0] = /* true */1;
    var l = nt.left;
    var r = nt.right;
    if (l !== null) {
      if (r !== null) {
        nt.right = Bs_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Bs_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else if (r !== null) {
      return r;
    } else {
      return l;
    }
  }
}

function removeCheck(d, v) {
  var dict = d.dict;
  var oldRoot = d.data;
  if (oldRoot !== null) {
    var removed = [/* false */0];
    var newRoot = removeMutateCheckAux(oldRoot, v, removed, dict[/* cmp */0]);
    if (newRoot !== oldRoot) {
      d.data = newRoot;
    }
    return removed[0];
  } else {
    return /* false */0;
  }
}

function addMutateCheckAux(t, x, added, cmp) {
  if (t !== null) {
    var k = t.key;
    var c = cmp(x, k);
    if (c) {
      var l = t.left;
      var r = t.right;
      if (c < 0) {
        var ll = addMutateCheckAux(l, x, added, cmp);
        t.left = ll;
      } else {
        t.right = addMutateCheckAux(r, x, added, cmp);
      }
      return Bs_internalAVLset.balMutate(t);
    } else {
      return t;
    }
  } else {
    added[0] = /* true */1;
    return Bs_internalAVLset.singleton0(x);
  }
}

function addCheck(m, e) {
  var dict = m.dict;
  var oldRoot = m.data;
  var added = [/* false */0];
  var newRoot = addMutateCheckAux(oldRoot, e, added, dict[/* cmp */0]);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
  }
  return added[0];
}

function split(d, key) {
  var dict = d.dict;
  var s = d.data;
  var arr = Bs_internalAVLset.toArray0(s);
  var i = Bs_Sort.binSearch(arr, key, dict[/* cmp */0]);
  var len = arr.length;
  if (i < 0) {
    var next = (-i | 0) - 1 | 0;
    return /* tuple */[
            /* tuple */[
              {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(arr, 0, next)
              },
              {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(arr, next, len - next | 0)
              }
            ],
            /* false */0
          ];
  } else {
    return /* tuple */[
            /* tuple */[
              {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(arr, 0, i)
              },
              {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(arr, i + 1 | 0, (len - i | 0) - 1 | 0)
              }
            ],
            /* true */1
          ];
  }
}

function keepBy(d, p) {
  var data = d.data;
  var dict = d.dict;
  return {
          dict: dict,
          data: Bs_internalAVLset.filterCopy(data, p)
        };
}

function partition(d, p) {
  var data = d.data;
  var dict = d.dict;
  var match = Bs_internalAVLset.partitionCopy(data, p);
  return /* tuple */[
          {
            dict: dict,
            data: match[0]
          },
          {
            dict: dict,
            data: match[1]
          }
        ];
}

function empty(dict) {
  return {
          dict: dict,
          data: Bs_internalAVLset.empty0
        };
}

function isEmpty(d) {
  return Bs_internalAVLset.isEmpty0(d.data);
}

function singleton(x, dict) {
  return {
          dict: dict,
          data: Bs_internalAVLset.singleton0(x)
        };
}

function minimum(d) {
  return Bs_internalAVLset.minOpt0(d.data);
}

function minNull(d) {
  return Bs_internalAVLset.minNull0(d.data);
}

function maximum(d) {
  return Bs_internalAVLset.maxOpt0(d.data);
}

function maxNull(d) {
  return Bs_internalAVLset.maxNull0(d.data);
}

function forEach(d, f) {
  return Bs_internalAVLset.iter0(d.data, f);
}

function reduce(d, acc, cb) {
  return Bs_internalAVLset.fold0(d.data, acc, cb);
}

function every(d, p) {
  return Bs_internalAVLset.every0(d.data, p);
}

function some(d, p) {
  return Bs_internalAVLset.some0(d.data, p);
}

function size(d) {
  return Bs_internalAVLset.length0(d.data);
}

function toList(d) {
  return Bs_internalAVLset.toList0(d.data);
}

function toArray(d) {
  return Bs_internalAVLset.toArray0(d.data);
}

function ofSortedArrayUnsafe(xs, dict) {
  return {
          dict: dict,
          data: Bs_internalAVLset.ofSortedArrayUnsafe0(xs)
        };
}

function checkInvariant(d) {
  return Bs_internalAVLset.checkInvariant(d.data);
}

function cmp(d0, d1) {
  var M = d0.dict;
  return Bs_internalAVLset.cmp0(d0.data, d1.data, M[/* cmp */0]);
}

function eq(d0, d1) {
  var M = d0.dict;
  return Bs_internalAVLset.eq0(d0.data, d1.data, M[/* cmp */0]);
}

function get(d, x) {
  var M = d.dict;
  return Bs_internalAVLset.findOpt0(d.data, x, M[/* cmp */0]);
}

function getNull(d, x) {
  var M = d.dict;
  return Bs_internalAVLset.findNull0(d.data, x, M[/* cmp */0]);
}

function getExn(d, x) {
  var dict = d.dict;
  return Bs_internalAVLset.findExn0(d.data, x, dict[/* cmp */0]);
}

function has(d, x) {
  var dict = d.dict;
  return Bs_internalAVLset.mem0(d.data, x, dict[/* cmp */0]);
}

function ofArray(data, dict) {
  return {
          dict: dict,
          data: Bs_internalAVLset.ofArray0(data, dict[/* cmp */0])
        };
}

function add(m, e) {
  var oldRoot = m.data;
  var M = m.dict;
  var newRoot = Bs_internalAVLset.addMutate(M[/* cmp */0], oldRoot, e);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
    return /* () */0;
  } else {
    return 0;
  }
}

function addArrayMutate(t, xs, cmp) {
  var v = t;
  for(var i = 0 ,i_finish = xs.length - 1 | 0; i <= i_finish; ++i){
    v = Bs_internalAVLset.addMutate(cmp, v, xs[i]);
  }
  return v;
}

function mergeMany(d, xs) {
  var oldRoot = d.data;
  var M = d.dict;
  var newRoot = addArrayMutate(oldRoot, xs, M[/* cmp */0]);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return /* () */0;
  } else {
    return 0;
  }
}

function subset(a, b) {
  var M = a.dict;
  return Bs_internalAVLset.subset0(a.data, b.data, M[/* cmp */0]);
}

function intersect(a, b) {
  var dict = a.dict;
  var dataa = a.data;
  var datab = b.data;
  if (dataa !== null) {
    if (datab !== null) {
      var sizea = Bs_internalAVLset.lengthNode(dataa);
      var sizeb = Bs_internalAVLset.lengthNode(datab);
      var totalSize = sizea + sizeb | 0;
      var tmp = new Array(totalSize);
      Bs_internalAVLset.fillArray(dataa, 0, tmp);
      Bs_internalAVLset.fillArray(datab, sizea, tmp);
      var p = dict[/* cmp */0];
      if (p(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || p(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
        return {
                dict: dict,
                data: Bs_internalAVLset.empty0
              };
      } else {
        var tmp2 = new Array(sizea < sizeb ? sizea : sizeb);
        var k = Bs_Sort.inter(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, p);
        return {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(tmp2, 0, k)
              };
      }
    } else {
      return {
              dict: dict,
              data: Bs_internalAVLset.empty0
            };
    }
  } else {
    return {
            dict: dict,
            data: Bs_internalAVLset.empty0
          };
  }
}

function diff(a, b) {
  var dict = a.dict;
  var dataa = a.data;
  var datab = b.data;
  if (dataa !== null) {
    if (datab !== null) {
      var sizea = Bs_internalAVLset.lengthNode(dataa);
      var sizeb = Bs_internalAVLset.lengthNode(datab);
      var totalSize = sizea + sizeb | 0;
      var tmp = new Array(totalSize);
      Bs_internalAVLset.fillArray(dataa, 0, tmp);
      Bs_internalAVLset.fillArray(datab, sizea, tmp);
      var p = dict[/* cmp */0];
      if (p(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || p(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
        return {
                dict: dict,
                data: Bs_internalAVLset.copy(dataa)
              };
      } else {
        var tmp2 = new Array(sizea);
        var k = Bs_Sort.diff(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, p);
        return {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(tmp2, 0, k)
              };
      }
    } else {
      return {
              dict: dict,
              data: Bs_internalAVLset.copy(dataa)
            };
    }
  } else {
    return {
            dict: dict,
            data: Bs_internalAVLset.empty0
          };
  }
}

function union(a, b) {
  var dict = a.dict;
  var dataa = a.data;
  var datab = b.data;
  if (dataa !== null) {
    if (datab !== null) {
      var sizea = Bs_internalAVLset.lengthNode(dataa);
      var sizeb = Bs_internalAVLset.lengthNode(datab);
      var totalSize = sizea + sizeb | 0;
      var tmp = new Array(totalSize);
      Bs_internalAVLset.fillArray(dataa, 0, tmp);
      Bs_internalAVLset.fillArray(datab, sizea, tmp);
      var p = dict[/* cmp */0];
      if (p(tmp[sizea - 1 | 0], tmp[sizea]) < 0) {
        return {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(tmp, 0, totalSize)
              };
      } else {
        var tmp2 = new Array(totalSize);
        var k = Bs_Sort.union(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, p);
        return {
                dict: dict,
                data: Bs_internalAVLset.ofSortedArrayAux(tmp2, 0, k)
              };
      }
    } else {
      return {
              dict: dict,
              data: Bs_internalAVLset.copy(dataa)
            };
    }
  } else {
    return {
            dict: dict,
            data: Bs_internalAVLset.copy(datab)
          };
  }
}

function copy(d) {
  return {
          dict: d.dict,
          data: Bs_internalAVLset.copy(d.data)
        };
}

exports.empty = empty;
exports.singleton = singleton;
exports.ofArray = ofArray;
exports.ofSortedArrayUnsafe = ofSortedArrayUnsafe;
exports.copy = copy;
exports.isEmpty = isEmpty;
exports.has = has;
exports.add = add;
exports.addCheck = addCheck;
exports.mergeMany = mergeMany;
exports.remove = remove;
exports.removeCheck = removeCheck;
exports.removeMany = removeMany;
exports.union = union;
exports.intersect = intersect;
exports.diff = diff;
exports.subset = subset;
exports.cmp = cmp;
exports.eq = eq;
exports.forEach = forEach;
exports.reduce = reduce;
exports.every = every;
exports.some = some;
exports.keepBy = keepBy;
exports.partition = partition;
exports.size = size;
exports.toList = toList;
exports.toArray = toArray;
exports.minimum = minimum;
exports.minNull = minNull;
exports.maximum = maximum;
exports.maxNull = maxNull;
exports.get = get;
exports.getNull = getNull;
exports.getExn = getExn;
exports.split = split;
exports.checkInvariant = checkInvariant;
/* No side effect */

File: jscomp/others/bs_SetM.ml

(* Copyright (C) 2017 Authors of BuckleScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)


module N = Bs_internalAVLset
module A = Bs_Array
module S = Bs_Sort 

type ('k,'id) t0 = 'k  N.t

type ('k, 'id) dict = ('k, 'id) Bs_Cmp.t 

type ('elt,'id) t =
  {
    dict: ('elt, 'id) dict;
    mutable data: 'elt N.t
  } [@@bs.deriving abstract]



let rec removeMutateAux nt x ~cmp = 
  let k = N.key nt in 
  let c = (Bs_Cmp.getCmpIntenral cmp) x k [@bs] in 
  if c = 0 then 
    let l,r = N.(left nt, right nt) in       
    match N.(toOpt l, toOpt r) with 
    | Some _,  Some nr ->  
      N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
      N.return (N.balMutate nt)
    | None, Some _ ->
      r  
    | (Some _ | None ), None ->  l 
  else 
    begin 
      if c < 0 then 
        match N.toOpt (N.left nt) with         
        | None -> N.return nt 
        | Some l ->
          N.leftSet nt (removeMutateAux ~cmp l x );
          N.return (N.balMutate nt)
      else 
        match N.toOpt (N.right nt) with 
        | None -> N.return nt 
        | Some r -> 
          N.rightSet nt (removeMutateAux ~cmp r x);
          N.return (N.balMutate nt)
    end

let remove (type elt) (type id) (d : (elt,id) t) v =  
  let oldRoot = data d in 
  let module M = (val dict d) in 
  match N.toOpt oldRoot with 
  | None -> ()
  | Some oldRoot2 ->
    let newRoot = removeMutateAux ~cmp:M.cmp oldRoot2 v in 
    if newRoot != oldRoot then 
      dataSet d newRoot    


let rec removeArrayMutateAux t xs i len ~cmp  =  
  if i < len then 
    let ele = A.getUnsafe xs i in 
    let u = removeMutateAux t ele ~cmp in 
    match N.toOpt u with 
    | None -> N.empty
    | Some t -> removeArrayMutateAux t xs (i+1) len ~cmp 
  else N.return t    

let removeMany (type elt) (type id) (d : (elt,id) t) xs =  
  let oldRoot = data d in 
  match N.toOpt oldRoot with 
  | None -> ()
  | Some nt -> 
    let len = A.length xs in 
    let module M = (val dict d) in 
    let newRoot = removeArrayMutateAux nt xs 0 len ~cmp:M.cmp in 
    if newRoot != oldRoot then 
      dataSet d newRoot


let rec removeMutateCheckAux  nt x removed ~cmp= 
  let k = N.key nt in 
  let c = (Bs_Cmp.getCmpIntenral cmp) x k [@bs] in 
  if c = 0 then 
    let () = removed := true in  
    let l,r = N.(left nt, right nt) in       
    match N.(toOpt l, toOpt r) with 
    | Some _,  Some nr ->  
      N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
      N.return (N.balMutate nt)
    | None, Some _ ->
      r  
    | (Some _ | None ), None ->  l 
  else 
    begin 
      if c < 0 then 
        match N.toOpt (N.left nt) with         
        | None -> N.return nt 
        | Some l ->
          N.leftSet nt (removeMutateCheckAux ~cmp l x removed);
          N.return (N.balMutate nt)
      else 
        match N.toOpt (N.right nt) with 
        | None -> N.return nt 
        | Some r -> 
          N.rightSet nt (removeMutateCheckAux ~cmp r x removed);
          N.return (N.balMutate nt)
    end



let removeCheck (type elt) (type id) (d : (elt,id) t) v =  
  let oldRoot = data d in 
  match N.toOpt oldRoot with 
  | None -> false 
  | Some oldRoot2 ->
    let module M = (val dict d) in
    let removed = ref false in 
    let newRoot = removeMutateCheckAux ~cmp:M.cmp oldRoot2 v removed in 
    if newRoot != oldRoot then  
      dataSet d newRoot ;   
    !removed



let rec addMutateCheckAux  (t : _ t0) x added ~cmp  =   
  match N.toOpt t with 
  | None -> 
    added := true;
    N.singleton x 
  | Some nt -> 
    let k = N.key nt in 
    let  c = (Bs_Cmp.getCmpIntenral cmp) x k [@bs] in  
    if c = 0 then t 
    else
      let l, r = N.(left nt, right nt) in 
      (if c < 0 then                   
         let ll = addMutateCheckAux ~cmp l x added in
         N.leftSet nt ll
       else   
         N.rightSet nt (addMutateCheckAux ~cmp r x added );
      );
      N.return (N.balMutate nt)

let addCheck (type elt) (type id) (m : (elt,id) t) e = 
  let dict, oldRoot = (dict m, data m) in 
  let module M = (val dict) in 
  let added = ref false in 
  let newRoot = addMutateCheckAux ~cmp:M.cmp oldRoot e added in 
  if newRoot != oldRoot then 
    dataSet m newRoot;
  !added    


let split (type elt) (type id) (d : (elt,id) t)  key  =     
  let module M = (val dict d ) in
  let arr = N.toArray (data d) in 
  let i = S.binarySearchBy arr key (Bs_Cmp.getCmpIntenral M.cmp)  in   
  let len = A.length arr in 
  if i < 0 then 
    let next = - i -1 in 
    (t 
       ~data:(N.ofSortedArrayAux arr 0 next)
       ~dict:(module M)
     , 
     t 
       ~data:(N.ofSortedArrayAux arr next (len - next))
       ~dict:(module M)
    ), false
  else 
    (t 
       ~data:(N.ofSortedArrayAux arr 0 i)
       ~dict:(module M),
     t 
       ~data:(N.ofSortedArrayAux arr (i+1) (len - i - 1))
       ~dict:(module M)
    ), true       

let keepBy d p = 
  t ~data:(N.filterCopy (data d) p ) ~dict:(dict d) 
let partition d p = 
  let dict = dict d in 
  let a, b = N.partitionCopy (data d) p in 
  t ~data:a ~dict, t ~data:b ~dict      

let empty ~dict = 
  t ~dict ~data:N.empty
    
let isEmpty d = 
  N.isEmpty (data d)
    
let singleton x ~dict = 
  t ~data:(N.singleton x) ~dict 
let minimum d = 
  N.minimum (data d)
let minUndefined d =
  N.minUndefined (data d)
let maximum d = 
  N.maximum (data d)
let maxUndefined d =
  N.maxUndefined (data d)
let forEach d f =
  N.forEach (data d) f     
let reduce d acc cb = 
  N.reduce (data d) acc cb 
let every d p = 
  N.every (data d) p 
let some d  p = 
  N.some (data d) p   
let size d = 
  N.size (data d)
let toList d =
  N.toList (data d)
let toArray d = 
  N.toArray (data d)
let ofSortedArrayUnsafe xs ~dict : _ t =
  t ~data:(N.ofSortedArrayUnsafe xs) ~dict   
let checkInvariantInternal d = 
  N.checkInvariantInternal (data d)
let cmp (type elt) (type id) (d0 : (elt,id) t) d1 = 
  let module M = (val dict d0) in 
  N.cmp ~cmp:M.cmp (data d0) (data d1)
let eq (type elt) (type id) (d0 : (elt,id) t)  d1 = 
  let module M = (val dict d0) in 
  N.eq ~cmp:M.cmp (data d0) (data d1)
let get (type elt) (type id) (d : (elt,id) t) x = 
  let module M = (val dict d) in 
  N.get ~cmp:M.cmp (data d) x 
let getUndefined (type elt) (type id) (d : (elt,id) t) x = 
  let module M = (val dict d) in 
  N.getUndefined ~cmp:M.cmp (data d) x
let getExn (type elt) (type id) (d : (elt,id) t) x = 
  let module M = (val dict d) in 
  N.getExn ~cmp:M.cmp (data d) x     
let has (type elt) (type id) (d : (elt,id) t) x =
  let module M = (val dict d) in 
  N.has ~cmp:M.cmp (data d) x   
let ofArray (type elt) (type id) data ~(dict : (elt,id) dict) =  
  let module M = (val dict) in 
  t ~dict ~data:(N.ofArray ~cmp:M.cmp data)
let add (type elt) (type id) (m : (elt,id) t) e = 
  let module M = (val dict m) in
  let oldRoot = (data m) in 
  let newRoot = N.addMutate ~cmp:M.cmp oldRoot e  in 
  if newRoot != oldRoot then 
    dataSet m newRoot

let addArrayMutate (t : _ t0) xs ~cmp =     
  let v = ref t in 
  for i = 0 to A.length xs - 1 do 
    v := N.addMutate !v (A.getUnsafe xs i)  ~cmp
  done; 
  !v
    
let mergeMany (type elt) (type id) (d : (elt,id) t ) xs =   
  let module M = (val dict d) in
  dataSet d (addArrayMutate (data d) xs ~cmp:M.cmp)







let subset (type elt) (type id) (a : (elt,id) t) b = 
  let module M = (val dict a) in 
  N.subset  ~cmp:M.cmp (data a) (data b)

let intersect (type elt) (type id) (a : (elt,id) t) b  : _ t = 
  (* let dict, dataa, datab = dict a, data a, data b in  *)
  let module M = (val dict a) in 
  match N.toOpt (data a), N.toOpt (data b) with 
  | None, _ -> empty (module M)
  | _, None -> empty (module M)
  | Some dataa0, Some datab0 ->  
    let sizea, sizeb = 
      N.lengthNode dataa0, N.lengthNode datab0 in          
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ; 
    ignore @@ N.fillArray datab0 sizea tmp;
    let p = Bs_Cmp.getCmpIntenral M.cmp in 
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       || 
       (p 
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0 
       )
    then empty (module M)
    else 
      let tmp2 = A.makeUninitializedUnsafe (min sizea sizeb) in 
      let k = S.intersect tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      t ~data:(N.ofSortedArrayAux tmp2 0 k)
        ~dict:(module M)
        
let diff (type elt) (type id) (a : (elt,id) t) b : _ t = 
  let module M = (val dict a) in
  let dataa = data a in 
  match N.toOpt dataa, N.toOpt (data b) with 
  | None, _ -> empty (module M)
  | _, None -> 
    t ~data:(N.copy dataa) ~dict:(module M) 
  | Some dataa0, Some datab0
    -> 
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in  
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ; 
    ignore @@ N.fillArray datab0 sizea tmp;
    let p = Bs_Cmp.getCmpIntenral M.cmp in 
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       || 
       (p 
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0 
       )
    then t ~data:(N.copy dataa) ~dict:(module M) 
    else 
      let tmp2 = A.makeUninitializedUnsafe sizea in 
      let k = S.diff tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      t ~data:(N.ofSortedArrayAux tmp2 0 k) ~dict:(module M) 

let union (type elt) (type id) (a : (elt,id) t) b = 
  let module M = (val dict a) in
  let dataa, datab =  data a, data b  in 
  match N.toOpt dataa, N.toOpt datab with 
  | None, _ -> t ~data:(N.copy datab) ~dict:(module M) 
  | _, None -> t ~data:(N.copy dataa) ~dict:(module M) 
  | Some dataa0, Some datab0 
    -> 
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in 
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ;
    ignore @@ N.fillArray datab0 sizea tmp ;
    let p = (Bs_Cmp.getCmpIntenral M.cmp)  in 
    if p
        (A.getUnsafe tmp (sizea - 1))
        (A.getUnsafe tmp sizea) [@bs] < 0 then 
      t ~data:(N.ofSortedArrayAux tmp 0 totalSize) ~dict:(module M) 
    else   
      let tmp2 = A.makeUninitializedUnsafe totalSize in 
      let k = S.union tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      t ~data:(N.ofSortedArrayAux tmp2 0 k) ~dict:(module M) 

let copy d = t ~data:(N.copy (data d)) ~dict:(dict d)
